validator/src/main/java/com/amazon/aoc/validators/CWMetricValidator.java

```diff
@@ -152,50 +153,57 @@ private void addMetrics(
       throws BaseException {
 
-    // load metrics into a hash set
-    Set<Metric> metricSet =
-        new TreeSet<>(
-            (Metric o1, Metric o2) -> {
-              // check namespace
-              if (!o1.getNamespace().equals(o2.getNamespace())) {
-                return o1.getNamespace().compareTo(o2.getNamespace());
-              }
-
-              // check metric name
-              if (!o1.getMetricName().equals(o2.getMetricName())) {
-                return o1.getMetricName().compareTo(o2.getMetricName());
-              }
-
-              // sort and check dimensions
-              List<Dimension> dimensionList1 = o1.getDimensions();
-              List<Dimension> dimensionList2 = o2.getDimensions();
-
-              // sort
-              dimensionList1.sort(Comparator.comparing(Dimension::getName));
-              dimensionList2.sort(Comparator.comparing(Dimension::getName));
-
-              return dimensionList1.toString().compareTo(dimensionList2.toString());
-            });
-    for (Metric metric : baseMetricList) {
-      metricSet.add(metric);
-    }
-    for (Metric metric : toBeCheckedMetricList) {
-      if (!metricSet.contains(metric)) {
-        throw new BaseException(
-            ExceptionCode.EXPECTED_METRIC_NOT_FOUND,
-            String.format(
-                "metric in %ntoBeCheckedMetricList: %s is not found in %nbaseMetricList: %s %n",
-                metric, metricSet));
+      Set<Metric> matchAny = new HashSet<>();
+      Set<Metric> matchExact = new HashSet<>();
+      for (Metric metric : expectedMetricList) {
+          metric.getDimensions().sort(Comparator.comparing(Dimension::getName));
+          if (metric.getDimensions().stream().anyMatch(d -> ANY_VALUE.equals(d.getValue()))) {
+              matchAny.add(metric);
+          } else {
+              matchExact.add(metric);
+          }
       }
-    }
+
+      Set<Metric> actualMetricSet = new HashSet<>();
+      for (Metric metric : actualMetricList) {
+         metric.getDimensions().sort(Comparator.comparing(Dimension::getName));
+         actualMetricSet.add(metric);
+      }
+      matchExact.removeAll(actualMetricSet);
+      if (!matchExact.isEmpty())  {
+          throw new BaseException(
+                  ExceptionCode.EXPECTED_METRIC_NOT_FOUND,
+                  String.format(
+                          "metric in %ntoBeCheckedMetricList: %s is not found in %nbaseMetricList: %s %n",
+                          matchExact.stream().findAny().get(), actualMetricSet));
+      }
+
+      Iterator<Metric> iter = matchAny.iterator();
+      while (iter.hasNext()) {
+          Metric expected = iter.next();
+          // There's a waste of computation because part of the actualMetricSet has been visited in the matchExact. Assuming
+          // the actualMetricSet size is small, do the comparison again to keep the logic simple.
+          for (Metric actual : actualMetricSet) {
```

Comment:
I wonder if there's a way to remove from the actual metric set the same metrics that were removed from `matchExact` during `matchExact.removeAll(actualMetricSet)` since they're already been matched.

This would reduce the number of metrics we iterate through drastically. Not a big issue though since we don't expect a huge number of metrics anyways.

Reply:
Done.